cli/node/mod.rs

```diff
@@ -0,0 +1,697 @@
+// Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
+
+use std::collections::HashSet;
+use std::path::Path;
+use std::path::PathBuf;
+
+use deno_ast::MediaType;
+use deno_ast::ModuleSpecifier;
+use deno_core::anyhow::Context;
+use deno_core::error::generic_error;
+use deno_core::error::AnyError;
+use deno_core::located_script_name;
+use deno_core::serde_json::Map;
+use deno_core::serde_json::Value;
+use deno_core::url::Url;
+use deno_core::JsRuntime;
+use deno_graph::source::ResolveResponse;
+use deno_runtime::deno_node::get_package_scope_config;
+use deno_runtime::deno_node::legacy_main_resolve;
+use deno_runtime::deno_node::package_exports_resolve;
+use deno_runtime::deno_node::package_imports_resolve;
+use deno_runtime::deno_node::package_resolve;
+use deno_runtime::deno_node::DenoDirNpmResolver;
+use deno_runtime::deno_node::PackageJson;
+use deno_runtime::deno_node::DEFAULT_CONDITIONS;
+use once_cell::sync::Lazy;
+use path_clean::PathClean;
+use regex::Regex;
+
+use crate::compat;
+use crate::file_fetcher::FileFetcher;
+use crate::npm::GlobalNpmPackageResolver;
+use crate::npm::NpmPackageReference;
+use crate::npm::NpmPackageResolver;
+
+mod analyze;
+
+pub use analyze::esm_code_with_node_globals;
+
+static RESERVED_WORDS: Lazy<HashSet<&str>> = Lazy::new(|| {
+  HashSet::from([
+    "break",
+    "case",
+    "catch",
+    "class",
+    "const",
+    "continue",
+    "debugger",
+    "default",
+    "delete",
+    "do",
+    "else",
+    "export",
+    "extends",
+    "false",
+    "finally",
+    "for",
+    "function",
+    "if",
+    "import",
+    "in",
+    "instanceof",
+    "new",
+    "null",
+    "return",
+    "super",
+    "switch",
+    "this",
+    "throw",
+    "true",
+    "try",
+    "typeof",
+    "var",
+    "void",
+    "while",
+    "with",
+    "yield",
+    "let",
+    "enum",
+    "implements",
+    "interface",
+    "package",
+    "private",
+    "protected",
+    "public",
+    "static",
+  ])
+});
+
+pub async fn initialize_runtime(
+  js_runtime: &mut JsRuntime,
+) -> Result<(), AnyError> {
+  let source_code = &format!(
+    r#"(async function loadBuiltinNodeModules(moduleAllUrl) {{
+      const moduleAll = await import(moduleAllUrl);
```

Comment:
It's unfortunate that we have to load all of the node modules even if only some are used. Do you think in the future these could be loaded lazily?

Reply:
I'd push back on that assumption. I'm sure there are many npm modules that don't require, say, crypto.